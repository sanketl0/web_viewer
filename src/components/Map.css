/* Force maplibre to fill all available space */
.gvs-map.maplibregl-map {
	width: 100%;
	height: 100%;
	background-color: white;
}

/* Picture thumbnail on map */
.maplibregl-popup-content {
	padding: 5px !important;
	position: relative;
}

.gvs-map-thumb {
	display: inline-block;
	text-align: center;
	vertical-align: middle;
	max-width: 130px;
	max-height: 130px;
}

@keyframes rotating {
	from { transform: rotate(0deg); }
	to { transform: rotate(360deg); }
}

.gvs-map-thumb-loader {
	background-color: rgb(230,230,230);
	border-radius: 65px;
	max-height: 60px;
	margin: 5px 35px;
	animation: rotating 2s linear infinite;
}

.gvs-map-thumb-legend {
	display: block;
	position: absolute;
	bottom: 5px;
	right: 5px;
	left: 5px;
	background-color: rgba(0,0,0,0.8);
	color: white;
	text-align: center;
	font-style: italic;
	font-size: 0.8em;
	margin: 0;
	padding: 0;
}

/* Max size for geocoder search bar */
.maplibregl-ctrl-geocoder {
	max-width: 60%;
}

.maplibregl-marker {
	width: 60px;
}

/* Resize canvas for print */
@media print {
	.gvs-map.maplibregl-map canvas {
		width: 100% !important;
		height: 100% !important;
	}
}
