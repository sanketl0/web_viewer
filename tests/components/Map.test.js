import Map from "../../src/components/Map";

jest.mock("maplibre-gl", () => ({
	addProtocol: jest.fn(),
	AttributionControl: jest.fn(),
	GeolocateControl: class {
		onAdd() {;}
	},
	Marker: jest.fn(),
	Popup: class {
		on() {;}
	},
	Map: class {
		constructor(opts) {
			this._mapOpts = opts;
		}
		getContainer() {
			return this._mapOpts.container;
		}
		addControl() {;}
		addSource() {;}
		addLayer() {;}
		getLayer() {;}
		setLayoutProperty() {;}
		getStyle() {
			return {
				layers: [],
				sources: {},
				metadata: {},
			};
		}
		resize() {;}
		on(type, handler) {
			if(!this._handlers) { this._handlers = {}; }
			if(!this._handlers[type]) { this._handlers[type] = []; }
			this._handlers[type].push(handler);
		}
		_fire(type) {
			this._handlers[type].forEach(f => f());
		}
	},
}));

const createParent = () => ({
	addEventListener: jest.fn(),
	dispatchEvent: jest.fn(),
	isWidthSmall: jest.fn(),
	_options: {
		users: ["geovisio"],
	},
	_api: {
		onceReady: () => Promise.resolve(),
		getDataBbox: jest.fn(),
		getPicturesTilesUrl: jest.fn(),
		_getMapRequestTransform: jest.fn(),
		getMapStyle: () => ({ sources: {}, layers: [], metadata: {} }),
	},
	_t: {
		maplibre: {},
	}
});


describe("reloadVectorTiles", () => {
	it("works", () => {
		const p = createParent();
		const c = document.createElement("div");
		const m = new Map(p, c);
		const setter = jest.fn();
		m.getSource = (user) => ({ tiles: [`https://bla.xyz/${user}/x/y/z`], setTiles: setter });
		m._userLayers = ["geovisio", "toto"];
		m.reloadVectorTiles();
		expect(setter.mock.calls).toMatchSnapshot();
	});
});

describe("hasTwoBackgrounds", () => {
	it("is true if 2 bg", () => {
		const p = createParent();
		const c = document.createElement("div");
		const m = new Map(p, c);
		m.getLayer = (id) => id == "gvs-aerial";
		expect(m.hasTwoBackgrounds()).toBeTruthy();
	});

	it("is false if 1 bg", () => {
		const p = createParent();
		const c = document.createElement("div");
		const m = new Map(p, c);
		m.getLayer = (id) => id == "gvs-aerial" ? undefined : {};
		expect(m.hasTwoBackgrounds()).toBeFalsy();
	});
});

describe("getBackground", () => {
	it("works if raster is enabled", () => {
		const p = createParent();
		const c = document.createElement("div");
		const m = new Map(p, c, { raster: { type: "raster" } });
		m.getLayer = () => true;
		m.getLayoutProperty = () => "hidden";
		expect(m.getBackground()).toBe("streets");
		m.getLayoutProperty = () => "visible";
		expect(m.getBackground()).toBe("aerial");
	});

	it("works if no raster enabled", () => {
		const p = createParent();
		const c = document.createElement("div");
		const m = new Map(p, c);
		expect(m.getBackground()).toBe("streets");
	});
});

describe("setBackground", () => {
	it("works if raster is enabled", () => {
		const p = createParent();
		p.dispatchEvent = jest.fn();
		const c = document.createElement("div");
		const m = new Map(p, c, { raster: { type: "raster" } });
		m.setLayoutProperty = jest.fn();
		m.getLayer = () => true;
		m.setBackground("aerial");
		expect(m.setLayoutProperty.mock.calls).toMatchSnapshot();
		expect(p.dispatchEvent.mock.calls).toMatchSnapshot();
	});

	it("skips if setting streets and no raster available", () => {
		const p = createParent();
		const c = document.createElement("div");
		const m = new Map(p, c);
		m.setLayoutProperty = jest.fn();
		m.setBackground("streets");
		expect(m.setLayoutProperty.mock.calls.length).toBe(0);
	});

	it("fails if no raster available", () => {
		const p = createParent();
		const c = document.createElement("div");
		const m = new Map(p, c);
		expect(() => m.setBackground("aerial")).toThrowError("No aerial imagery available");
	});
});

describe("getVisibleUsers", () => {
	it("works", async () => {
		const p = createParent();
		const c = document.createElement("div");
		const m = new Map(p, c);
		m.getSource = () => true;
		m.setPaintProperty = jest.fn();
		await m._postLoad();
		m.getLayoutProperty = () => "visible";
		expect(m.getVisibleUsers()).toStrictEqual(["geovisio"]);
	});
});

describe("setVisibleUsers", () => {
	it("works when no users exists", async () => {
		const p = createParent();
		const c = document.createElement("div");
		const m = new Map(p, c);
		m._createPicturesTilesLayer = (url, id) => m._userLayers.add(id);
		m.setLayoutProperty = jest.fn();
		await m.setVisibleUsers(["blabla"]);
		expect(m.setLayoutProperty.mock.calls).toMatchSnapshot();
		expect(p.dispatchEvent.mock.calls).toMatchSnapshot();
	});

	it("works when user already exist but is hidden", async () => {
		const p = createParent();
		p._options.users = ["blabla", "geovisio"];
		const c = document.createElement("div");
		const m = new Map(p, c);
		m.setPaintProperty = jest.fn();
		m.getSource = () => true;
		m.getLayer = () => true;
		let cptlCount = 0;
		let deCalls = [];
		return new Promise(async (resolve) => {
			p.dispatchEvent = (e) => {
				deCalls.push(e);
				if(e.type == "map:users-changed") {
					resolve();
				}
			};
			await m._postLoad();
		}).then(() => {
			m.setLayoutProperty = jest.fn();
			m._createPicturesTilesLayer = () => { cptlCount++; return Promise.resolve(); };
			expect(m._userLayers).toEqual(new Set(["blabla", "geovisio"]));
			m.setVisibleUsers(["blabla"]);
		}).then(() => {
			expect(cptlCount).toBe(0);
			expect(deCalls).toMatchSnapshot();
			expect(m.setLayoutProperty.mock.calls).toMatchSnapshot();
		});
	});
});

describe("filterUserLayersContent", () => {
	it("works", async () => {
		const p = createParent();
		p._options.users = ["blabla", "geovisio"];
		const c = document.createElement("div");
		const m = new Map(p, c);
		m.getSource = () => true;
		m.setPaintProperty = jest.fn();
		m.getLayer = () => true;
		await m._postLoad();
		m.setFilter = jest.fn();
		m.filterUserLayersContent("pictures", [["test", "true"]]);
		expect(m.setFilter.mock.calls).toMatchSnapshot();
	});
});

describe("reloadLayersStyles", () => {
	it("works", async () => {
		const p = createParent();
		p._options.users = ["blabla", "geovisio"];
		const c = document.createElement("div");
		const m = new Map(p, c);
		m.getSource = () => true;
		m.setPaintProperty = jest.fn();
		await m._postLoad();
		m.setLayoutProperty = jest.fn();
		m.setPaintProperty = jest.fn();
		m.reloadLayersStyles();
		expect(m.setLayoutProperty.mock.calls).toMatchSnapshot();
		expect(m.setPaintProperty.mock.calls).toMatchSnapshot();
	});
});
