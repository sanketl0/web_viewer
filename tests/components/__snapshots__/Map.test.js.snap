// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`filterUserLayersContent works 1`] = `
Array [
  Array [
    "geovisio_blabla_pictures",
    Array [
      Array [
        "test",
        "true",
      ],
    ],
  ],
  Array [
    "geovisio_blabla_pictures_symbols",
    Array [
      Array [
        "test",
        "true",
      ],
    ],
  ],
  Array [
    "geovisio_pictures",
    Array [
      Array [
        "test",
        "true",
      ],
    ],
  ],
  Array [
    "geovisio_pictures_symbols",
    Array [
      Array [
        "test",
        "true",
      ],
    ],
  ],
]
`;

exports[`reloadLayersStyles works 1`] = `
Array [
  Array [
    "geovisio_blabla_pictures",
    "circle-sort-key",
    Array [
      "case",
      Array [
        "==",
        Array [
          "get",
          "hidden",
        ],
        true,
      ],
      90,
      10,
    ],
  ],
  Array [
    "geovisio_pictures",
    "circle-sort-key",
    Array [
      "case",
      Array [
        "==",
        Array [
          "get",
          "hidden",
        ],
        true,
      ],
      90,
      10,
    ],
  ],
  Array [
    "geovisio_blabla_pictures_symbols",
    "icon-image",
    Array [
      "case",
      Array [
        "==",
        Array [
          "get",
          "id",
        ],
        undefined,
      ],
      "",
      Array [
        "==",
        Array [
          "get",
          "type",
        ],
        "equirectangular",
      ],
      "gvs-arrow-360",
      "gvs-arrow-flat",
    ],
  ],
  Array [
    "geovisio_blabla_pictures_symbols",
    "symbol-sort-key",
    Array [
      "case",
      Array [
        "==",
        Array [
          "get",
          "hidden",
        ],
        true,
      ],
      90,
      10,
    ],
  ],
  Array [
    "geovisio_pictures_symbols",
    "icon-image",
    Array [
      "case",
      Array [
        "==",
        Array [
          "get",
          "id",
        ],
        undefined,
      ],
      "",
      Array [
        "==",
        Array [
          "get",
          "type",
        ],
        "equirectangular",
      ],
      "gvs-arrow-360",
      "gvs-arrow-flat",
    ],
  ],
  Array [
    "geovisio_pictures_symbols",
    "symbol-sort-key",
    Array [
      "case",
      Array [
        "==",
        Array [
          "get",
          "hidden",
        ],
        true,
      ],
      90,
      10,
    ],
  ],
  Array [
    "geovisio_blabla_sequences",
    "line-sort-key",
    Array [
      "case",
      Array [
        "==",
        Array [
          "get",
          "hidden",
        ],
        true,
      ],
      90,
      10,
    ],
  ],
  Array [
    "geovisio_blabla_sequences",
    "line-cap",
    "square",
  ],
  Array [
    "geovisio_sequences",
    "line-sort-key",
    Array [
      "case",
      Array [
        "==",
        Array [
          "get",
          "hidden",
        ],
        true,
      ],
      90,
      10,
    ],
  ],
  Array [
    "geovisio_sequences",
    "line-cap",
    "square",
  ],
]
`;

exports[`reloadLayersStyles works 2`] = `
Array [
  Array [
    "geovisio_blabla_pictures",
    "circle-color",
    Array [
      "case",
      Array [
        "==",
        Array [
          "get",
          "hidden",
        ],
        true,
      ],
      "#34495E",
      Array [
        "==",
        Array [
          "get",
          "geovisio:status",
        ],
        "hidden",
      ],
      "#34495E",
      "#FF6F00",
    ],
  ],
  Array [
    "geovisio_blabla_pictures",
    "circle-radius",
    Array [
      "interpolate",
      Array [
        "linear",
      ],
      Array [
        "zoom",
      ],
      15,
      4.5,
      18,
      6,
      24,
      12,
    ],
  ],
  Array [
    "geovisio_blabla_pictures",
    "circle-opacity",
    Array [
      "interpolate",
      Array [
        "linear",
      ],
      Array [
        "zoom",
      ],
      15,
      0,
      16,
      1,
    ],
  ],
  Array [
    "geovisio_blabla_pictures",
    "circle-stroke-color",
    "#ffffff",
  ],
  Array [
    "geovisio_blabla_pictures",
    "circle-stroke-width",
    Array [
      "interpolate",
      Array [
        "linear",
      ],
      Array [
        "zoom",
      ],
      16,
      0,
      17,
      1,
      18,
      1.5,
      24,
      3,
    ],
  ],
  Array [
    "geovisio_pictures",
    "circle-color",
    Array [
      "case",
      Array [
        "==",
        Array [
          "get",
          "hidden",
        ],
        true,
      ],
      "#34495E",
      Array [
        "==",
        Array [
          "get",
          "geovisio:status",
        ],
        "hidden",
      ],
      "#34495E",
      "#FF6F00",
    ],
  ],
  Array [
    "geovisio_pictures",
    "circle-radius",
    Array [
      "interpolate",
      Array [
        "linear",
      ],
      Array [
        "zoom",
      ],
      15,
      4.5,
      18,
      6,
      24,
      12,
    ],
  ],
  Array [
    "geovisio_pictures",
    "circle-opacity",
    Array [
      "interpolate",
      Array [
        "linear",
      ],
      Array [
        "zoom",
      ],
      15,
      0,
      16,
      1,
    ],
  ],
  Array [
    "geovisio_pictures",
    "circle-stroke-color",
    "#ffffff",
  ],
  Array [
    "geovisio_pictures",
    "circle-stroke-width",
    Array [
      "interpolate",
      Array [
        "linear",
      ],
      Array [
        "zoom",
      ],
      16,
      0,
      17,
      1,
      18,
      1.5,
      24,
      3,
    ],
  ],
  Array [
    "geovisio_blabla_sequences",
    "line-color",
    Array [
      "case",
      Array [
        "==",
        Array [
          "get",
          "hidden",
        ],
        true,
      ],
      "#34495E",
      Array [
        "==",
        Array [
          "get",
          "geovisio:status",
        ],
        "hidden",
      ],
      "#34495E",
      "#FF6F00",
    ],
  ],
  Array [
    "geovisio_blabla_sequences",
    "line-width",
    Array [
      "interpolate",
      Array [
        "linear",
      ],
      Array [
        "zoom",
      ],
      0,
      0.5,
      10,
      2,
      14,
      4,
      16,
      5,
      22,
      3,
    ],
  ],
  Array [
    "geovisio_sequences",
    "line-color",
    Array [
      "case",
      Array [
        "==",
        Array [
          "get",
          "hidden",
        ],
        true,
      ],
      "#34495E",
      Array [
        "==",
        Array [
          "get",
          "geovisio:status",
        ],
        "hidden",
      ],
      "#34495E",
      "#FF6F00",
    ],
  ],
  Array [
    "geovisio_sequences",
    "line-width",
    Array [
      "interpolate",
      Array [
        "linear",
      ],
      Array [
        "zoom",
      ],
      0,
      0.5,
      10,
      2,
      14,
      4,
      16,
      5,
      22,
      3,
    ],
  ],
]
`;

exports[`reloadVectorTiles works 1`] = `
Array [
  Array [
    Array [
      "https://bla.xyz/geovisio/x/y/z",
    ],
  ],
  Array [
    Array [
      "https://bla.xyz/geovisio_toto/x/y/z",
    ],
  ],
]
`;

exports[`setBackground works if raster is enabled 1`] = `
Array [
  Array [
    "gvs-aerial",
    "visibility",
    "visible",
  ],
]
`;

exports[`setBackground works if raster is enabled 2`] = `
Array [
  Array [
    CustomEvent {
      "isTrusted": false,
    },
  ],
]
`;

exports[`setVisibleUsers works when no users exists 1`] = `Array []`;

exports[`setVisibleUsers works when no users exists 2`] = `
Array [
  Array [
    CustomEvent {
      "isTrusted": false,
    },
  ],
]
`;

exports[`setVisibleUsers works when user already exist but is hidden 1`] = `
Array [
  CustomEvent {
    "isTrusted": false,
  },
  CustomEvent {
    "isTrusted": false,
  },
]
`;

exports[`setVisibleUsers works when user already exist but is hidden 2`] = `
Array [
  Array [
    "geovisio_blabla_pictures",
    "circle-sort-key",
    Array [
      "case",
      Array [
        "==",
        Array [
          "get",
          "hidden",
        ],
        true,
      ],
      90,
      10,
    ],
  ],
  Array [
    "geovisio_pictures",
    "circle-sort-key",
    Array [
      "case",
      Array [
        "==",
        Array [
          "get",
          "hidden",
        ],
        true,
      ],
      90,
      10,
    ],
  ],
  Array [
    "geovisio_blabla_pictures_symbols",
    "icon-image",
    Array [
      "case",
      Array [
        "==",
        Array [
          "get",
          "id",
        ],
        undefined,
      ],
      "",
      Array [
        "==",
        Array [
          "get",
          "type",
        ],
        "equirectangular",
      ],
      "gvs-arrow-360",
      "gvs-arrow-flat",
    ],
  ],
  Array [
    "geovisio_blabla_pictures_symbols",
    "symbol-sort-key",
    Array [
      "case",
      Array [
        "==",
        Array [
          "get",
          "hidden",
        ],
        true,
      ],
      90,
      10,
    ],
  ],
  Array [
    "geovisio_pictures_symbols",
    "icon-image",
    Array [
      "case",
      Array [
        "==",
        Array [
          "get",
          "id",
        ],
        undefined,
      ],
      "",
      Array [
        "==",
        Array [
          "get",
          "type",
        ],
        "equirectangular",
      ],
      "gvs-arrow-360",
      "gvs-arrow-flat",
    ],
  ],
  Array [
    "geovisio_pictures_symbols",
    "symbol-sort-key",
    Array [
      "case",
      Array [
        "==",
        Array [
          "get",
          "hidden",
        ],
        true,
      ],
      90,
      10,
    ],
  ],
  Array [
    "geovisio_blabla_sequences",
    "line-sort-key",
    Array [
      "case",
      Array [
        "==",
        Array [
          "get",
          "hidden",
        ],
        true,
      ],
      90,
      10,
    ],
  ],
  Array [
    "geovisio_blabla_sequences",
    "line-cap",
    "square",
  ],
  Array [
    "geovisio_sequences",
    "line-sort-key",
    Array [
      "case",
      Array [
        "==",
        Array [
          "get",
          "hidden",
        ],
        true,
      ],
      90,
      10,
    ],
  ],
  Array [
    "geovisio_sequences",
    "line-cap",
    "square",
  ],
  Array [
    "geovisio_blabla_pictures",
    "visibility",
    "visible",
  ],
  Array [
    "geovisio_blabla_sequences",
    "visibility",
    "visible",
  ],
  Array [
    "geovisio_blabla_sequences_plus",
    "visibility",
    "visible",
  ],
  Array [
    "geovisio_blabla_grid",
    "visibility",
    "visible",
  ],
  Array [
    "geovisio_blabla_pictures_symbols",
    "visibility",
    "visible",
  ],
  Array [
    "geovisio_pictures",
    "visibility",
    "none",
  ],
  Array [
    "geovisio_sequences",
    "visibility",
    "none",
  ],
  Array [
    "geovisio_sequences_plus",
    "visibility",
    "none",
  ],
  Array [
    "geovisio_grid",
    "visibility",
    "none",
  ],
  Array [
    "geovisio_pictures_symbols",
    "visibility",
    "none",
  ],
]
`;
